### PR TITLE
Feature/filter post taxonomies component output

### PR DESCRIPTION
**Pull request by dsawardekar**
## Description
This PR adds the filter~~s `editor.FlatTermSelector` and `editor.HierarchicalTermSelector`~~ `editor.PostTaxonomyType`. They allow developers the ability to customize the Taxonomy Selector UI for custom taxonomies. 

The specific use case I'm attempting to solve is to allow a way to limit Taxonomy selection to a single term, by changing the UI to use a Dropdown.

Similar to #6940. /cc @adamsilverstein 

## How has this been tested?

With Gutenberg on this feature branch, I added the following filter to my plugin JS.

```js
var el = wp.element.createElement;

function customizeProductTypeSelector( OriginalComponent ) {
	return function( props ) {
		if ( props.slug === 'category' ) {
			return el(
				'div',
				{},
				'Product Type Selector'
			);
		} else {
			return el(
				OriginalComponent,
				props
			);
		}
	}
};

wp.hooks.addFilter(
	'editor.PostTaxonomyType',
	'my-custom-plugin',
	customizeProductTypeSelector
);
```

I was able to customize the UI as seen in the Screenshot.

## Screenshots 

Custom taxonomy selector

## Types of changes

New Feature to allow customization of Post Taxonomy Selector(s) in the sidebar.

## Checklist:
- [x ] My code is tested.
- [x ] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by chrisvanpatten** [comment]
I know the Yoast folks need this too for their "Primary Category" feature (I could have sworn they had posted an issue here about it, but I can't find it).

Excited about this!

**Comment by dsawardekar** [comment]
I'm not sure why the Preview Window test is failing. The message suggests a connection error. Can we rerun the tests?

```log
 FAIL  test/e2e/specs/preview.test.js (5.253s)
  ● Preview › Should open a preview window for a new post
    Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined
      at Promise (node_modules/puppeteer/lib/Connection.js:198:56)
      at CDPSession.send (node_modules/puppeteer/lib/Connection.js:197:12)
      at ExecutionContext.evaluateHandle (node_modules/puppeteer/lib/ExecutionContext.js:71:75)
      at ExecutionContext.evaluate (node_modules/puppeteer/lib/ExecutionContext.js:46:31)
      at Frame.evaluate (node_modules/puppeteer/lib/FrameManager.js:299:20)
```

File: editor/components/post-taxonomies/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+Similarly, to modify the contents of the HierarchicalTermSelector component use the hook `editor.HierarchicalTermSelector`,
+
+For example, to render alternative UI for the taxonomy `product-category`,
+we can render custom markup or use the Original component as shown below.
+
+```js
+const { addFilter } = wp.hooks;
+
+function CustomHierarchicalTermSelector(OriginalHierarchicalTermSelector) {
+	return (props) => {
```
Comment (adamsilverstein) [inline editor/components/post-taxonomies/README.md]: code example could use whitespace to match coding standards

File: editor/components/post-taxonomies/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+Similarly, to modify the contents of the HierarchicalTermSelector component use the hook `editor.HierarchicalTermSelector`,
+
+For example, to render alternative UI for the taxonomy `product-category`,
+we can render custom markup or use the Original component as shown below.
+
+```js
+const { addFilter } = wp.hooks;
+
+function CustomHierarchicalTermSelector(OriginalHierarchicalTermSelector) {
```
Comment (adamsilverstein) [inline editor/components/post-taxonomies/README.md]: needs whitespace

File: editor/components/post-taxonomies/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	return (props) => {
+		if ( props.slug === 'product-category' ) {
+			return (
+				<div>Custom HierarchicalTermSelector Term Selector</div>
+			);
+		} else {
+			return <OriginalHierarchicalTermSelector {...props} />
```
Comment (adamsilverstein) [inline editor/components/post-taxonomies/README.md]: whitespace around `...props`

File: editor/components/post-taxonomies/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+const { addFilter } = wp.hooks;
+
+function CustomFlatTermSelector(OriginalFlatTermSelector) {
+	return (props) => {
+		if ( props.slug === 'product-type' ) {
+			return (
+				<div>Custom Flat Term Selector</div>
+			);
+		} else {
+			return <OriginalFlatTermSelector {...props} />
```
Comment (adamsilverstein) [inline editor/components/post-taxonomies/README.md]: whitespace

File: editor/components/post-taxonomies/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+To modify the contents of the FlatTermSelector component use the hook `editor.FlatTermSelector`,
+
+For example, to render alternative UI for the taxonomy `product-type`,
+we can render custom markup or use the Original component as shown below.
+
+```js
+const { addFilter } = wp.hooks;
+
+function CustomFlatTermSelector(OriginalFlatTermSelector) {
+	return (props) => {
```
Comment (adamsilverstein) [inline editor/components/post-taxonomies/README.md]: whitespace

**Review by adamsilverstein — DISMISSED** [review]
A few whitespace changes in the docs please, then I think this is good to go as long as the team agrees its an appropriate extensibility point.

**Comment by dsawardekar** [comment]
Thanks, @adamsilverstein I've updated those code examples.

**Review by gziolo — COMMENTED** [review]
In general, this is looking really good. I'm wondering if we could use only one filter instead of two proposed which would work almost exactly the same.

It might be tricky when I look at the current implementation of `PostTaxonomies`:
https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/post-taxonomies/index.js#L19-L35

I'm thinking about something along those lines:
```js


function customizeHierarchicalTermSelector( OriginalComponent ) {
	if ( OriginalComponent !== HierarchicalTermSelector ) {
		return OriginalComponent;
	}
	return ( props ) => {
		if (  props.slug === 'product-type' ) {
			return (
				<div>Custom Hierarchical Term Selector</div>
			);
		} else {
			return <OriginalComponent { ...props } />
		}
	}
};

addFilter(
	'editor.PostTaxonomies',
	'my-custom-plugin',
	customizeHierarchicalTermSelector
);
```

It's not ideal. My point is, that it would be nicer to have only one filter in here which would cover both use cases.

**Comment by krisgale** [comment]
have a similar need that would be served well by this... i seek to be able to add a way to allow an editor to choose one of the Categories as a primary category (likely via a Make Primary hyperlink, as seen here: https://yoast.com/gutenberg-integration/ and here: https://wordpress.org/plugins/easy-primary-category/ ), and at present, have no good way to introduce additinoal, subtle functionality to Categories.

**Comment by dsawardekar** [comment]
Thanks, @gziolo. The `editor.PostTaxonomies` filter would be cleaner. I've updated this PR to reflect.

**Comment by gziolo** [comment]
I spent some more time trying to better understand how the code is structured. It looks like `PostTaxonomies` represents multiple taxonomy panels:
screen shot 2018-08-01 at 08 19 07

Do you think it would make sense to introduce a proxy component which would allow passing the information whether it is hierarchical? This would require turning:
```js
const TaxonomyComponent = taxonomy.hierarchical ? HierarchicalTermSelector : FlatTermSelector;
```

into its own component and apply this filter there.

In code it would be something like:

```jsx
function PostTaxonomiesItem( { hierarchical, slug ) {
    if ( hierarchical ) {
        return <HierarchicalTermSelector slug={ slug } />;
    }

    return <FlatTermSelector slug={ slug } />;
}

export default withFilters(
    'editor.PostTaxonomiesItem'
)(
    PostTaxonomiesItem
);
```

Please note, that I used **singular** form for the filter and component name.

**Comment by dsawardekar** [comment]
@gziolo The extra Taxonomy information would be useful to have, but not having it doesn't prevent customization.

From the Product Type example, once you have the Taxonomy slug, you already have adequate information to render the custom UI.

**Comment by gziolo** [comment]
I updated code example to use ES5 syntax, so everyone could copy and paste it into their browser.

**Review by gziolo — APPROVED** [review]

**Pull request merged**